subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaLibraryPlugin.java

```diff
@@ -0,0 +1,88 @@
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.api.internal.artifacts.configurations.ConfigurationInternal;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.compile.JavaCompile;
+
+import static org.gradle.api.plugins.JavaPlugin.USAGE;
+
+/**
+ * <p>A {@link Plugin} which extends the capabilities of the {@link JavaPlugin Java plugin} by cleanly separating
+ * the API and implementation dependencies of a library.</p>
+ */
+public class JavaLibraryPlugin implements Plugin<Project> {
+
+    @Override
+    public void apply(final Project project) {
+        project.getPluginManager().apply(JavaPlugin.class);
+
+        JavaPluginConvention convention = (JavaPluginConvention) project.getConvention().getPlugins().get("java");
+        final ConfigurationContainer configurations = project.getConfigurations();
+        addApiToMainSourceSet(project, convention, configurations);
+    }
+
+    private void addApiToMainSourceSet(final Project project, JavaPluginConvention convention, final ConfigurationContainer configurations) {
+        convention.getSourceSets().getByName("main", new Action<SourceSet>() {
+            @Override
+            public void execute(SourceSet sourceSet) {
+                defineApiConfigurationsForSourceSet(sourceSet, configurations);
+                JavaCompile javaCompile = (JavaCompile) project.getTasks().findByName(sourceSet.getCompileJavaTaskName());
+                project.getArtifacts().add(sourceSet.getApiCompileConfigurationName(), ImmutableMap.of(
+                    "file", javaCompile.getDestinationDir(),
+                    "builtBy", javaCompile));
+            }
+        });
+    }
+
+    private void defineApiConfigurationsForSourceSet(SourceSet sourceSet, ConfigurationContainer configurations) {
+        Configuration apiConfiguration = configurations.maybeCreate(sourceSet.getApiConfigurationName());
+        apiConfiguration.setVisible(false);
+        apiConfiguration.setDescription("API dependencies for " + sourceSet + ".");
+        apiConfiguration.setCanBeResolved(false);
+        apiConfiguration.setCanBeConsumed(false);
+        apiConfiguration.setTransitive(false);
+
+        Configuration apiCompileConfiguration = configurations.maybeCreate(sourceSet.getApiCompileConfigurationName());
+        apiCompileConfiguration.setVisible(false);
+        apiCompileConfiguration.setDescription("API compile classpath for " + sourceSet + ".");
+        apiCompileConfiguration.setCanBeResolved(true);
+        apiCompileConfiguration.setCanBeConsumed(true);
+        apiCompileConfiguration.setTransitive(false);
+        apiCompileConfiguration.attribute(USAGE, "api");
+        apiCompileConfiguration.extendsFrom(apiConfiguration);
+
+        Configuration implementationConfiguration = configurations.maybeCreate(sourceSet.getImplementationConfigurationName());
```

Comment:
Well I asked on Slack and since I didn't get a comment I went ahead with _not_ adding it to base to avoid confusion. I don't think we need to replace `compile` in the end for `java`. The `java` plugin would typically be used for applications, and whether we use `compile` or `implementation` here as the name for dependencies isn't really relevant. However, having suddenly a new configuration for users of the `java` plugin who do not care (yet) about api/impl separation is IMHO confusing. It's only used for libraries, so I kept it there.